# Data Description 


Estimating churners prior to their abandonment of a product or service is crucial. In this machine learning assignment, I will design a churn prediction model for the telecom industry to forecast which customers are most likely to churn.

The 39 features of this dataset are as follows:

1. Acronyms: Descriptions
2. MOBILE_NUMBER: Customer phone number
3. CIRCLE_ID: Telecom circle area to which the customer belongs to
4. LOC	Local calls: within same telecom circle
5. STD	STD calls: outside the calling circle
6. IC: Incoming calls
7. OG: Outgoing calls
8. T2T: Operator T to T, i.e. within same operator (mobile to mobile)
9. T2M: Operator T to other operator mobile
10. T2O: Operator T to other operator fixed line
11. T2F: Operator T to fixed lines of T
12. T2C: Operator T to it’s own call center
13. ARPU: Average revenue per user
14. MOU: Minutes of usage - voice calls
15. AON: Age on network - number of days the customer is using the operator T network
16. ONNET: All kind of calls within the same operator network
17. OFFNET:    	All kind of calls outside the operator T network
18. ROAM:	Indicates that customer is in roaming zone during the call
19. SPL:   	Special calls
20. ISD:    	ISD calls
21. RECH:    	Recharge
22. NUM:    	Number
23. AMT:    	Amount in local currency
24. MAX:    	Maximum
25. DATA:    	Mobile internet
26. 3G:    	3G network
27. AV:    	Average
28. VOL:    	Mobile internet usage volume (in MB)
29. 2G:    	2G network
30. PCK:    	Prepaid service schemes called - PACKS
31. NIGHT:    	Scheme to use during specific night hours only
32. MONTHLY:    	Service schemes with validity equivalent to a month
33. SACHET:   	Service schemes with validity smaller than a month
34. *.6:    	KPI for the month of June
35. *.7:    	KPI for the month of July
36. *.8:    	KPI for the month of August
37. *.9:    	KPI for the month of September
38. FB_USER:	Service scheme to avail services of Facebook and similar social networking sites
39. VBC:    	Volume based cost - when no specific scheme is not purchased and paid as per usage

# Problem Statement

In the telecommunications business, clients may pick from a variety of service providers and actively switch from one to another. In this extremely competitive sector, the yearly turnover rate for the telecoms industry averages between 15 and 25 percent. Given that it costs five to ten times as much to gain a new client as it does to maintain an existing one, customer retention is now more crucial than customer acquisition.

Retaining high-profitable clients is the primary business objective for many incumbent operators of T. To decrease customer turnover, telecom businesses must anticipate which consumers are at high risk of churning.

# Aim

In this project, I am going to analyze the customer-level data of a large telecommunications company, develop predictive models to identify consumers at high risk of churn, and identify the primary signs of churn.


# Tech stack

* Language – Python
* Libraries - Numpy, Pandas, Matplotlib, Seaborn, Scikit-learn,

# Approach

1. Importing the required libraries and reading the dataset.
    * Understanding the dataset
2. Exploratory Data Analysis (EDA) –
3. Filtering High Value Customers
4. Creating target Variable
5. Developing new features
6. Handling Missing values
7. Data Visualization-Univariate Analysis
8. Data Visualization- Bivariate Analysis
9. Outlier Detection
10. Data Preparation
11. Data Modeling and Eavlaution
12. Non-Interpretable Models
13. Interpretable Models
14. Conclusion

# 1. Understanding the dataset

In [1]:
%%HTML
<style type="text/css">
table.dataframe td, table.dataframe th {
    border: 1px  black solid !important;
  color: black !important;
}
</style>

## 1.1 Importing Libraries

In [2]:
!pip install graphviz

In [3]:
#Importing Data Reading and Processing Libraries
import pandas as pd
import numpy as np

#Imporitng Data Visualization Libraries
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

#Importing Data Preparation and Modeling Libraries
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold, GridSearchCV,StratifiedKFold

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier


#Importing Warning Libraries
import warnings
warnings.filterwarnings("ignore")

#Importing Miscellaneous Libraries
pd.set_option("display.max_columns",None)
pd.set_option("display.max_rows",None)
pd.set_option('display.width', None)

# Import the StandardScaler()
from sklearn.preprocessing import StandardScaler

#Improting the PCA module
from sklearn.decomposition import PCA

# For Hopkins test
from sklearn.neighbors import NearestNeighbors
from random import sample
from numpy.random import uniform
import numpy as np
from math import isnan

# For clustering 
## using KMeans ##
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

# Importing classification report and confusion matrix from sklearn metrics
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
from sklearn.metrics import recall_score,roc_auc_score,roc_curve

## using Hierarchical ##
from scipy.cluster.hierarchy import linkage
from scipy.cluster.hierarchy import dendrogram
from scipy.cluster.hierarchy import cut_tree

# Importing required packages for visualization
from IPython.display import Image  
#from sklearn.externals.six import StringIO  
from sklearn.tree import export_graphviz
import pydot, graphviz


# Other sklearn packages
import sklearn.metrics as metrics
from xgboost import XGBClassifier
from sklearn.tree import DecisionTreeClassifier

from datetime import date,datetime
import math
import multiprocessing

print("Successfully Importing Libraries...")

Successfully Importing Libraries...


# 1.2 Dataset Loading 

In [4]:
df = pd.read_csv('telecom_churn_data.csv')
print("Successully Datset Load..")
df.sample(2)

Successully Datset Load..


mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
26249     7000125802        109             0.0             0.0   
35840     7000140523        109             0.0             0.0   

       loc_ic_t2o_mou last_date_of_month_6 last_date_of_month_7  \
26249             0.0            6/30/2014            7/31/2014   
35840             0.0            6/30/2014            7/31/2014   

      last_date_of_month_8 last_date_of_month_9    arpu_6   arpu_7    arpu_8  \
26249            8/31/2014            9/30/2014   196.446    14.54   304.631   
35840            8/31/2014            9/30/2014  1690.836  2064.99  1850.739   

         arpu_9  onnet_mou_6  onnet_mou_7  onnet_mou_8  onnet_mou_9  \
26249    73.547         0.00         0.00        56.73         0.00   
35840  2052.249       450.53      1518.56      1457.88      2682.61   

       offnet_mou_6  offnet_mou_7  offnet_mou_8  offnet_mou_9  roam_ic_mou_6  \
26249         10.04         10.09        119.41          0.00            0.0   
35840       1990.31       2038.28       1794.41       1204.14            0.0   

       roam_ic_mou_7  roam_ic_mou_8  roam_ic_mou_9  roam_og_mou_6  \
26249            0.0            0.0            0.0            0.0   
35840            0.0            0.0            0.0            0.0   

       roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  loc_og_t2t_mou_6  \
26249            0.0            0.0            0.0              0.00   
35840            0.0            0.0            0.0            148.31   

       loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  loc_og_t2m_mou_6  \
26249              0.00              2.56              0.00              0.00   
35840           1074.36            548.34            217.08           1125.04   

       loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  loc_og_t2f_mou_6  \
26249              1.23              7.11              0.00               0.0   
35840           1657.83           1129.91            321.38               0.0   

       loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  loc_og_t2c_mou_6  \
26249              0.00              0.00               0.0              2.06   
35840              0.25              0.13               0.0              0.00   

       loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_t2c_mou_9  loc_og_mou_6  \
26249               0.0              0.00               0.0          0.00   
35840               0.0              2.66               0.0       1273.36   

       loc_og_mou_7  loc_og_mou_8  loc_og_mou_9  std_og_t2t_mou_6  \
26249          1.23          9.68          0.00              0.00   
35840       2732.44       1678.39        538.46            302.21   

       std_og_t2t_mou_7  std_og_t2t_mou_8  std_og_t2t_mou_9  std_og_t2m_mou_6  \
26249              0.00             54.16              0.00              0.00   
35840            444.19            909.53           2465.53            865.09   

       std_og_t2m_mou_7  std_og_t2m_mou_8  std_og_t2m_mou_9  std_og_t2f_mou_6  \
26249              7.06             93.59              0.00               0.0   
35840            380.19            660.23            882.76               0.0   

       std_og_t2f_mou_7  std_og_t2f_mou_8  std_og_t2f_mou_9  std_og_t2c_mou_6  \
26249               1.8               0.0               0.0               0.0   
35840               0.0               0.0               0.0               0.0   

       std_og_t2c_mou_7  std_og_t2c_mou_8  std_og_t2c_mou_9  std_og_mou_6  \
26249               0.0               0.0               0.0          0.00   
35840               0.0               0.0               0.0       1167.31   

       std_og_mou_7  std_og_mou_8  std_og_mou_9  isd_og_mou_6  isd_og_mou_7  \
26249          8.86        147.76          0.00           0.0           0.0   
35840        824.39       1569.76       3348.29           0.0           0.0   

       isd_og_mou_8  isd_og_mou_9  spl_og_mou_6  spl_og_mou_7  spl_og_mou_8  \
26249           0.0          

In [5]:
df.shape

(99999, 226)

`Observation`: Number of rows: 99999, 
               Number of columns: 226 

In [6]:
df.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 226 columns):
 #    Column                    Dtype  
---   ------                    -----  
 0    mobile_number             int64  
 1    circle_id                 int64  
 2    loc_og_t2o_mou            float64
 3    std_og_t2o_mou            float64
 4    loc_ic_t2o_mou            float64
 5    last_date_of_month_6      object 
 6    last_date_of_month_7      object 
 7    last_date_of_month_8      object 
 8    last_date_of_month_9      object 
 9    arpu_6                    float64
 10   arpu_7                    float64
 11   arpu_8                    float64
 12   arpu_9                    float64
 13   onnet_mou_6               float64
 14   onnet_mou_7               float64
 15   onnet_mou_8               float64
 16   onnet_mou_9               float64
 17   offnet_mou_6              float64
 18   offnet_mou_7              float64
 19   offnet_mou_8              float64
 20   offn

In [7]:
#Let's examine the data's distribution.
df.describe().T

count          mean            std           min  \
mobile_number       99999.0  7.001207e+09  695669.386290  7.000000e+09   
circle_id           99999.0  1.090000e+02       0.000000  1.090000e+02   
loc_og_t2o_mou      98981.0  0.000000e+00       0.000000  0.000000e+00   
std_og_t2o_mou      98981.0  0.000000e+00       0.000000  0.000000e+00   
loc_ic_t2o_mou      98981.0  0.000000e+00       0.000000  0.000000e+00   
arpu_6              99999.0  2.829874e+02     328.439770 -2.258709e+03   
arpu_7              99999.0  2.785366e+02     338.156291 -2.014045e+03   
arpu_8              99999.0  2.791547e+02     344.474791 -9.458080e+02   
arpu_9              99999.0  2.616451e+02     341.998630 -1.899505e+03   
onnet_mou_6         96062.0  1.323959e+02     297.207406  0.000000e+00   
onnet_mou_7         96140.0  1.336708e+02     308.794148  0.000000e+00   
onnet_mou_8         94621.0  1.330181e+02     308.951589  0.000000e+00   
onnet_mou_9         92254.0  1.303023e+02     308.477668  0.000000e+00   
offnet_mou_6        96062.0  1.979356e+02     316.851613  0.000000e+00   
offnet_mou_7        96140.0  1.970451e+02     325.862803  0.000000e+00   
offnet_mou_8        94621.0  1.965748e+02     327.170662  0.000000e+00   
offnet_mou_9        92254.0  1.903372e+02     319.396092  0.000000e+00   
roam_ic_mou_6       96062.0  9.950013e+00      72.825411  0.000000e+00   
roam_ic_mou_7       96140.0  7.149898e+00      73.447948  0.000000e+00   
roam_ic_mou_8       94621.0  7.292981e+00      68.402466  0.000000e+00   
roam_ic_mou_9       92254.0  6.343841e+00      57.137537  0.000000e+00   
roam_og_mou_6       96062.0  1.391134e+01      71.443196  0.000000e+00   
roam_og_mou_7       96140.0  9.818732e+00      58.455762  0.000000e+00   
roam_og_mou_8       94621.0  9.971890e+00      64.713221  0.000000e+00   
roam_og_mou_9       92254.0  8.555519e+00      58.438186  0.000000e+00   
loc_og_t2t_mou_6    96062.0  4.710076e+01     150.856393  0.000000e+00   
loc_og_t2t_mou_7    96140.0  4.647301e+01     155.318705  0.000000e+00   
loc_og_t2t_mou_8    94621.0  4.588781e+01     151.184830  0.000000e+00   
loc_og_t2t_mou_9    92254.0  4.458445e+01     147.995390  0.000000e+00   
loc_og_t2m_mou_6    96062.0  9.334209e+01     162.780544  0.000000e+00   
loc_og_t2m_mou_7    96140.0  9.139713e+01     157.492308  0.000000e+00   
loc_og_t2m_mou_8    94621.0  9.175513e+01     156.537048  0.000000e+00   
loc_og_t2m_mou_9    92254.0  9.046319e+01     158.681454  0.000000e+00   
loc_og_t2f_mou_6    96062.0  3.751013e+00      14.230438  0.000000e+00   
loc_og_t2f_mou_7    96140.0  3.792985e+00      14.264986  0.000000e+00   
loc_og_t2f_mou_8    94621.0  3.677991e+00      13.270996  0.000000e+00   
loc_og_t2f_mou_9    92254.0  3.655123e+00      13.457549  0.000000e+00   
loc_og_t2c_mou_6    96062.0  1.123056e+00       5.448946  0.000000e+00   
loc_og_t2c_mou_7    96140.0  1.368500e+00       7.533445  0.000000e+00   
loc_og_t2c_mou_8    94621.0  1.433821e+00       6.783335  0.000000e+00   
loc_og_t2c_mou_9    92254.0  1.232726e+00       5.619021  0.000000e+00   
loc_og_mou_6        96062.0  1.442012e+02     251.751489  0.000000e+00   
loc_og_mou_7        96140.0  1.416705e+02     248.731086  0.000000e+00   
loc_og_mou_8        94621.0  1.413282e+02     245.914311  0.000000e+00   
loc_og_mou_9        92254.0  1.387100e+02     245.934517  0.000000e+00   
std_og_t2t_mou_6    96062.0  7.982987e+01     252.476533  0.000000e+00   
std_og_t2t_mou_7    96140.0  8.329960e+01     263.631042  0.000000e+00   
std_og_t2t_mou_8    94621.0  8.328267e+01     265.486090  0.000000e+00   
std_og_t2t_mou_9    92254.0  8.234292e+01     267.184991  0.000000e+00   
std_og_t2m_mou_6    96062.0  8.729962e+01     255.617850  0.000000e+00   
std_og_t2m_mou_7    96140.0  9.080414e+01     269.347911  0.000000e+00   
std_og_t2m_mou_8    94621.0  8.983839e+01     271.757783  0.000000e+00   
std_og_t2m_mou_9    92254.0  8.627662e+01     261.407396  0.000000e+00   
std_og_t2f_mou_6    96062.

# 2. Data Cleaning

Some columns representing volume-based users include the month in their names, whereas the remaining columns use the numerals "6,7,8" to indicate the month. I am planning to rename the missplead column.

In [8]:
month = ['aug_vbc_3g','jul_vbc_3g','jun_vbc_3g','sep_vbc_3g']
df = df.rename(columns = {'aug_vbc_3g':'vbc_3g_8','jul_vbc_3g':'vbc_3g_7','jun_vbc_3g':'vbc_3g_6',
                          'sep_vbc_3g':'vbc_3g_9'})


## 2.1 Converting Datetime

In [9]:
#The conversion of date columns to date-time format

date_col= [col for col in df.columns if 'date' in col]

for i in df[date_col]:
    df[i] = pd.to_datetime(df[i])


In [10]:
df.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 226 columns):
 #    Column                    Dtype         
---   ------                    -----         
 0    mobile_number             int64         
 1    circle_id                 int64         
 2    loc_og_t2o_mou            float64       
 3    std_og_t2o_mou            float64       
 4    loc_ic_t2o_mou            float64       
 5    last_date_of_month_6      datetime64[ns]
 6    last_date_of_month_7      datetime64[ns]
 7    last_date_of_month_8      datetime64[ns]
 8    last_date_of_month_9      datetime64[ns]
 9    arpu_6                    float64       
 10   arpu_7                    float64       
 11   arpu_8                    float64       
 12   arpu_9                    float64       
 13   onnet_mou_6               float64       
 14   onnet_mou_7               float64       
 15   onnet_mou_8               float64       
 16   onnet_mou_9               float64     

# 2.2 Removing single unique value
Columns with a single unique value for all clients are essentially meaningless. As a result, i am deleting columns with zero variance from the dataset.

In [11]:
cols = []
for i in df.columns:
    if df[i].nunique() ==1:
        cols.append(i)
df = df.drop(cols,axis=1)
df.sample(2)

mobile_number   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  \
70246     7000965690  422.317  338.046  476.039  426.563       228.34   
15289     7000391649  123.511   41.810   45.330   46.250        12.84   

       onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
70246       181.91       342.86       225.04        171.29        126.84   
15289         0.00         0.40         0.00        229.09         15.83   

       offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  \
70246        114.99        189.83            0.0           0.00   
15289         29.13         30.49            0.0          23.69   

       roam_ic_mou_8  roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  \
70246           0.00           0.00            0.0           0.00   
15289           8.18           4.88            0.0          15.83   

       roam_og_mou_8  roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  \
70246           0.00           0.00             77.18             63.94   
15289          29.53          30.49             12.84              0.00   

       loc_og_t2t_mou_8  loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  \
70246             72.68             60.23             53.94             42.23   
15289              0.00              0.00             11.51              0.00   

       loc_og_t2m_mou_8  loc_og_t2m_mou_9  loc_og_t2f_mou_6  loc_og_t2f_mou_7  \
70246             66.56             64.71               0.0               0.0   
15289              0.00              0.00               0.0               0.0   

       loc_og_t2f_mou_8  loc_og_t2f_mou_9  loc_og_t2c_mou_6  loc_og_t2c_mou_7  \
70246               0.0               0.0               0.0              4.36   
15289               0.0               0.0               0.0              0.00   

       loc_og_t2c_mou_8  loc_og_t2c_mou_9  loc_og_mou_6  loc_og_mou_7  \
70246              8.43              0.11        131.13        106.18   
15289              0.00              0.00         24.36          0.00   

       loc_og_mou_8  loc_og_mou_9  std_og_t2t_mou_6  std_og_t2t_mou_7  \
70246        139.24        124.94            151.16            117.96   
15289          0.00          0.00              0.00              0.00   

       std_og_t2t_mou_8  std_og_t2t_mou_9  std_og_t2m_mou_6  std_og_t2m_mou_7  \
70246            270.18            164.81            117.34             84.61   
15289              0.00              0.00            217.58              0.00   

       std_og_t2m_mou_8  std_og_t2m_mou_9  std_og_t2f_mou_6  std_og_t2f_mou_7  \
70246             48.43            125.11               0.0               0.0   
15289              0.00              0.00               0.0               0.0   

       std_og_t2f_mou_8  std_og_t2f_mou_9  std_og_mou_6  std_og_mou_7  \
70246               0.0               0.0        268.51        202.58   
15289               0.0               0.0        217.58          0.00   

       std_og_mou_8  std_og_mou_9  isd_og_mou_6  isd_og_mou_7  isd_og_mou_8  \
70246        318.61        289.93           0.0           0.0           0.0   
15289          0.00          0.00           0.0           0.0           0.0   

       isd_og_mou_9  spl_og_mou_6  spl_og_mou_7  spl_og_mou_8  spl_og_mou_9  \
70246           0.0           0.0          5.33         10.64          0.73   
15289           0.0           0.0          0.00          0.00          0.00   

       og_others_6  og_others_7  og_others_8  og_others_9  total_og_mou_6  \
70246          0.0          0.0          0.0          0.0          399.64   
15289          0.0          0.0          0.0          0.0          241.94   

       total_og_mou_7  total_og_mou_8  total_og_mou_9  loc_ic_t2t_mou_6  \
70246          314.09          468.51          415.61             42.04   
15289            0.00            0.00            0.00             14.58   

       loc_ic_t2t_mou_7  loc_ic_t2t_mou_8  loc_ic_t2t_mou_9  loc_ic_t2m_mou_6  \
70246             35.28    

In [12]:
df.shape

(99999, 210)

`Observation`: Number of columns: unique column 16 is dropped from 226 to 210. 

# 2.3 Create ID for each customer

Eliminate the mobile number column and create an id column for client identification.

In [13]:
#drop mobile number
df = df.drop('mobile_number', axis=1)

#Create id
df = df.reset_index()
df = df.rename(columns = {'index':'cust_id'})
df['cust_id'] = df['cust_id']+1
df.sample(2)

cust_id   arpu_6   arpu_7   arpu_8  arpu_9  onnet_mou_6  onnet_mou_7  \
53873    53874  105.358  120.532  151.441  174.51       182.88       114.21   
71196    71197  144.709  100.963   15.810    0.00       156.39       192.33   

       onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
53873       187.23       298.53         22.96         35.39         33.19   
71196         3.21         0.00          6.01          6.31          0.00   

       offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
53873         17.94           0.00            0.0           0.00   
71196          0.00           5.16            0.0           1.13   

       roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
53873            0.0           0.00            0.0           0.00   
71196            2.1          10.26            0.0           3.68   

       roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
53873            0.0               0.0              0.96              4.61   
71196            0.0               0.0              5.79              0.00   

       loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
53873              9.79              6.34             12.71             25.24   
71196              0.00              0.78              2.23              0.00   

       loc_og_t2m_mou_9  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
53873              8.78              0.00              0.71               0.0   
71196              0.00              0.15              1.63               0.0   

       loc_og_t2f_mou_9  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
53873              0.51              0.00              0.26               0.0   
71196              0.00              1.28              0.00               0.0   

       loc_og_t2c_mou_9  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  \
53873              0.06          6.34         14.39         29.86   
71196              0.00          0.93          9.66          0.00   

       loc_og_mou_9  std_og_t2t_mou_6  std_og_t2t_mou_7  std_og_t2t_mou_8  \
53873         19.09            182.88            113.24            182.61   
71196          0.00            146.13            186.53              0.00   

       std_og_t2t_mou_9  std_og_t2m_mou_6  std_og_t2m_mou_7  std_og_t2m_mou_8  \
53873            288.73             16.61             16.64              7.94   
71196              0.00              3.79              2.45              0.00   

       std_og_t2m_mou_9  std_og_t2f_mou_6  std_og_t2f_mou_7  std_og_t2f_mou_8  \
53873              7.93               0.0               0.0               0.0   
71196              0.00               0.0               0.0               0.0   

       std_og_t2f_mou_9  std_og_mou_6  std_og_mou_7  std_og_mou_8  \
53873               0.0        199.49        129.89        190.56   
71196               0.0        149.93        188.98          0.00   

       std_og_mou_9  isd_og_mou_6  isd_og_mou_7  isd_og_mou_8  isd_og_mou_9  \
53873        296.66           0.0           0.0           0.0           0.0   
71196          0.00           0.0           0.0           0.0           0.0   

       spl_og_mou_6  spl_og_mou_7  spl_og_mou_8  spl_og_mou_9  og_others_6  \
53873          0.00          5.98          0.73          4.83         0.43   
71196          1.28          0.76          0.76          0.00         0.00   

       og_others_7  og_others_8  og_others_9  total_og_mou_6  total_og_mou_7  \
53873          0.0          0.0          0.0          206.28          150.28   
71196          0.0          0.0          0.0          152.14          199.41   

       total_og_mou_8  total_og_mou_9  loc_ic_t2t_mou_6  loc_ic_t2t_mou_7  \
53873          221.16          320.59              1.88             12.28   
71196            0.76            0.00              3.91             13.34   

       loc_ic_t2t_mou_8  loc_ic_t2t_mou_9  loc_ic_t2m_mou_6  loc_ic_t2m_mou

## 2.4 Separate Categorical columns

Check if this dataset has any category columns. This may be determined by identifying which columns contain the numbers 1 and 0. These values are equivalent to yes and no, respectively.

In [14]:
cat_cols = []

for i in df.columns:
    if df[i].nunique()==2:
        cat_cols.append(i)
cat_cols 

['night_pck_user_6',
 'night_pck_user_7',
 'night_pck_user_8',
 'night_pck_user_9',
 'fb_user_6',
 'fb_user_7',
 'fb_user_8',
 'fb_user_9']

## 2.5 Some Important Insights

### 2.5.1 Avaerage Revenue of call per user

In [15]:
arpu=[]
arpu_col= [col for col in df.columns if 'arpu' in col]

for i in df[arpu_col]:
    x=df[[i]].min()
    arpu.append(x)
print(arpu)

[arpu_6   -2258.709
dtype: float64, arpu_7   -2014.045
dtype: float64, arpu_8   -945.808
dtype: float64, arpu_9   -1899.505
dtype: float64, arpu_3g_6   -30.82
dtype: float64, arpu_3g_7   -26.04
dtype: float64, arpu_3g_8   -24.49
dtype: float64, arpu_3g_9   -71.09
dtype: float64, arpu_2g_6   -35.83
dtype: float64, arpu_2g_7   -15.48
dtype: float64, arpu_2g_8   -55.83
dtype: float64, arpu_2g_9   -45.74
dtype: float64]


`Observation`
The minimum values in arpu 6, arpu 7, arpu 8, and arpu 9 are seen to be negative. This suggests that certain clients are causing a loss for the business. I will maintain them in my study since my research criteria for a high-value client are based on usage-based churn, not revenue-based churn. Getting rid of them might result in the loss of some insightful knowledge. Observe their significance in the exploratory data analysis section before making a decision.

### 2.5.2 Sachet recharge
Recharge sachets are Service plans having a duration of less than one month. This indicates that the number of days needed to recharge a sachet should be fewer than 30. Any service plans that extend beyond 29 days should indicate that the client has performed a monthly recharge or that the entry is incorrect. Let us limit the values beyond 29 days to the maximum number of days recharged below 30 days.

In [16]:
sachet=[]
sachet_col= [col for col in df.columns if 'sachet' in col]

for i in df[sachet_col]:
    x=df[[i]].max()
    sachet.append(x)
print(sachet)

[sachet_2g_6    42
dtype: int64, sachet_2g_7    48
dtype: int64, sachet_2g_8    44
dtype: int64, sachet_2g_9    40
dtype: int64, sachet_3g_6    29
dtype: int64, sachet_3g_7    35
dtype: int64, sachet_3g_8    41
dtype: int64, sachet_3g_9    49
dtype: int64]


In [17]:
df['sachet_2g_6'] = df['sachet_2g_6'].clip(0,28)
df['sachet_2g_7'] = df['sachet_2g_7'].clip(0,29)
df['sachet_2g_8'] = df['sachet_2g_8'].clip(0,29)
df['sachet_3g_6'] = df['sachet_3g_6'].clip(0,29)
df['sachet_3g_7'] = df['sachet_3g_7'].clip(0,24)
df['sachet_3g_8'] = df['sachet_3g_8'].clip(0,29)

In [18]:
df.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 210 columns):
 #    Column                    Dtype         
---   ------                    -----         
 0    cust_id                   int64         
 1    arpu_6                    float64       
 2    arpu_7                    float64       
 3    arpu_8                    float64       
 4    arpu_9                    float64       
 5    onnet_mou_6               float64       
 6    onnet_mou_7               float64       
 7    onnet_mou_8               float64       
 8    onnet_mou_9               float64       
 9    offnet_mou_6              float64       
 10   offnet_mou_7              float64       
 11   offnet_mou_8              float64       
 12   offnet_mou_9              float64       
 13   roam_ic_mou_6             float64       
 14   roam_ic_mou_7             float64       
 15   roam_ic_mou_8             float64       
 16   roam_ic_mou_9             float64     

# 3 Focusing on high value Customers

In [19]:
df['total_amt_6'] = df[['total_rech_amt_6', 'total_rech_data_6']].sum(axis=1)
df['total_amt_7'] = df[['total_rech_amt_7', 'total_rech_data_7']].sum(axis=1)
df['total_amt_8'] = df[['total_rech_amt_8', 'total_rech_data_8']].sum(axis=1)
df['total_amt_9'] = df[['total_rech_amt_9', 'total_rech_data_9']].sum(axis=1)

In [20]:
#calculate total recharge amount per user for months
df['total_rech_amt_per_user']=df[['total_amt_6','total_amt_7','total_amt_8','total_amt_9']].sum(axis=1)

In [21]:
df.sample(2)

cust_id   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  onnet_mou_7  \
8594      8595   87.837   87.300  751.817  256.749         0.00        13.48   
98308    98309  199.809  387.321  304.869   97.177       152.81       480.09   

       onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
8594          0.00         0.41         38.16         78.23          1.38   
98308       405.61        80.43        245.01        297.01        223.71   

       offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
8594          18.19            0.0            0.0            0.0   
98308         75.08            0.7            0.0            0.0   

       roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
8594             0.0           0.00            0.0            0.0   
98308            0.0           0.65            0.0            0.0   

       roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
8594             0.0              0.00             13.48              0.00   
98308            0.0              3.74             24.74             14.96   

       loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
8594               0.41              7.71             34.94              1.38   
98308              1.16             26.43             26.74             15.33   

       loc_og_t2m_mou_9  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
8594               5.64               0.0               0.0              0.00   
98308              6.64               0.0               1.7              3.18   

       loc_og_t2f_mou_9  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
8594                0.0              0.00              0.00               0.0   
98308               0.0              0.89              5.08               0.0   

       loc_og_t2c_mou_9  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  \
8594                0.0          7.71         48.43          1.38   
98308               0.0         30.18         53.19         33.48   

       loc_og_mou_9  std_og_t2t_mou_6  std_og_t2t_mou_7  std_og_t2t_mou_8  \
8594           6.06              0.00              0.00              0.00   
98308          7.81            148.41            455.34            390.64   

       std_og_t2t_mou_9  std_og_t2m_mou_6  std_og_t2m_mou_7  std_og_t2m_mou_8  \
8594               0.00             28.64             31.66              0.00   
98308             79.26            201.08            251.83            195.81   

       std_og_t2m_mou_9  std_og_t2f_mou_6  std_og_t2f_mou_7  std_og_t2f_mou_8  \
8594              12.54               0.0               0.0               0.0   
98308             54.34               0.0               0.0               0.0   

       std_og_t2f_mou_9  std_og_mou_6  std_og_mou_7  std_og_mou_8  \
8594                0.0         28.64         31.66          0.00   
98308               0.0        349.49        707.18        586.46   

       std_og_mou_9  isd_og_mou_6  isd_og_mou_7  isd_og_mou_8  isd_og_mou_9  \
8594          12.54           0.0           0.0           0.0           0.0   
98308        133.61           0.0           0.0           0.0           0.0   

       spl_og_mou_6  spl_og_mou_7  spl_og_mou_8  spl_og_mou_9  og_others_6  \
8594           1.80         11.61          0.00          0.00         0.00   
98308         19.09         18.03         10.24         14.48         4.81   

       og_others_7  og_others_8  og_others_9  total_og_mou_6  total_og_mou_7  \
8594           0.0          0.0          0.0           38.16           91.71   
98308          0.0          0.0          0.0          403.59          778.41   

       total_og_mou_8  total_og_mou_9  loc_ic_t2t_mou_6  loc_ic_t2t_mou_7  \
8594             1.38           18.61              0.00              2.29   
98308          630.19          155.91              4.06             31.13   

       loc_ic_t2t_mou_8  loc_ic_t2t_mou_9  loc_ic_t2m_mou_6  loc_ic_t2m_

## 3.1 Filtering high value customers

In [22]:
avg_amt_rech=df['total_rech_amt_per_user'].mean()

In [24]:
high_val_customer = df.loc[df['total_rech_amt_per_user'] >= avg_amt_rech]
high_val_customer.head()

cust_id    arpu_6    arpu_7    arpu_8   arpu_9  onnet_mou_6  onnet_mou_7  \
3         4   221.338   251.102   508.054  389.500        99.91        54.39   
7         8  1069.180  1349.850  3171.480  500.000        57.84        54.68   
8         9   378.721   492.223   137.362  166.787       413.69       351.03   
13       14   492.846   205.671   593.260  322.732       501.76       108.39   
15       16    31.000   510.465   590.643  510.390          NaN       246.56   

    onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
3        310.98       241.71        123.31        109.01         71.68   
7         52.29          NaN        453.43        567.16        325.91   
8         35.08        33.46         94.66         80.63        136.48   
13       534.24       244.81        413.31        119.28        482.46   
15       280.31       289.79           NaN        839.58       1011.91   

    offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  roam_ic_mou_9  \
3         113.54           0.00          54.86          44.38           0.00   
7            NaN          16.23          33.49          31.64            NaN   
8         108.71           0.00           0.00           0.00           0.00   
13        214.06          23.53         144.24          72.11         136.78   
15        642.14            NaN           0.00           0.00           0.00   

    roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  \
3            0.00          28.09          39.04           0.00   
7           23.74          12.59          38.06            NaN   
8            0.00           0.00           0.00           0.00   
13           7.98          35.26           1.44          12.78   
15            NaN           0.88           0.00           0.00   

    loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  \
3              73.68             34.81             10.61             15.49   
7              51.39             31.38             40.28               NaN   
8             297.13            217.59             12.49             26.13   
13             49.63              6.19             36.01              6.14   
15               NaN             16.13             44.79             48.33   

    loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  \
3             107.43             83.21             22.46             65.46   
7             308.63            447.38            162.28               NaN   
8              80.96             70.58             50.54             34.58   
13            151.13             47.28            294.46            108.24   
15               NaN             38.99             92.53            158.11   

    loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  \
3               1.91              0.65              4.91              2.06   
7              62.13             55.14             53.23               NaN   
8               0.00              0.00              0.00              0.00   
13              4.54              0.00             23.51              5.29   
15               NaN              5.13              4.83              8.43   

    loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_t2c_mou_9  \
3                0.0              0.00              0.00              0.00   
7                0.0              0.00              0.00               NaN   
8                0.0              0.00              7.15              0.00   
13               0.0              0.00              0.49              0.00   
15               NaN              9.78              0.00              0.03   

    loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  loc_og_mou_9  std_og_t2t_mou_6  \
3         183.03        118.68         37.99         83.03             26.23   
7         422.16        533.91        255.79           NaN              4.30   
8         378.09        288.18         63.04         60.71            116.56   
13  

In [27]:
print("Total number of high value customers: ",len(high_val_customer))

Total number of high value customers:  34958


# 4 Creating Target Variable¶

In this step, I will tag the churned customers whose target variable is categorised as churn Yes or churn No (Yes=1, No=0) based on T-four-month Mobile's data.


## 4.1 Find put inactive customers

During the churn period, those who have not made any calls (incoming or outgoing) and have not utilized mobile internet even once are considered to be inactive. The following attributes must be used to identify churners:

* total_ic_mou_6
* total_ic_mou_7
* total_ic_mou_8
* total_ic_mou_9
* total_og_mou_6
* total_og_mou_7
* total_og_mou_8
* total_og_mou_9
* vol_2g_mb_6
* vol_2g_mb_7
* vol_2g_mb_8
* vol_2g_mb_9
* vol_3g_mb_6
* vol_3g_mb_7
* vol_3g_mb_8
* vol_3g_mb_9

In [28]:
# The month of September is used to screen out churned customers.

df['churn'] = df.apply(lambda x: 1 if (x.total_ic_mou_9 == 0 and x.total_og_mou_9 == 0 and x.vol_2g_mb_9 ==0 and x.vol_3g_mb_9==0) else 0, axis=1)
df['churn'] = df['churn'].astype("str")
df.shape

(99999, 216)

In [29]:
# number of churned customers
df['churn'].value_counts()

0    89808
1    10191
Name: churn, dtype: int64

In [37]:
#what's the % of churned customers
print("The Percentage of churned customers is:" , round(100*(df.churn.astype("int").sum()/len(df)),2))

The Percentage of churned customers is: 10.19


### 4.2 After labeling churners, let's delete any characteristics pertaining to the churn phase (those with' 9', etc. in their names).

In [38]:
col_9 = [i for i in df.columns if '9' in i]
df = df.drop(col_9,axis=1)
df.shape

(99999, 163)

In [39]:
# let's update our categorical column list
cat_cols = [ele for ele in cat_cols if ele not in col_9]
cat_cols

['night_pck_user_6',
 'night_pck_user_7',
 'night_pck_user_8',
 'fb_user_6',
 'fb_user_7',
 'fb_user_8']

In [40]:
df.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 163 columns):
 #    Column                    Dtype         
---   ------                    -----         
 0    cust_id                   int64         
 1    arpu_6                    float64       
 2    arpu_7                    float64       
 3    arpu_8                    float64       
 4    onnet_mou_6               float64       
 5    onnet_mou_7               float64       
 6    onnet_mou_8               float64       
 7    offnet_mou_6              float64       
 8    offnet_mou_7              float64       
 9    offnet_mou_8              float64       
 10   roam_ic_mou_6             float64       
 11   roam_ic_mou_7             float64       
 12   roam_ic_mou_8             float64       
 13   roam_og_mou_6             float64       
 14   roam_og_mou_7             float64       
 15   roam_og_mou_8             float64       
 16   loc_og_t2t_mou_6          float64     

In [41]:
df.sample(2)

cust_id   arpu_6   arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  \
71941    71942   72.822  129.473  112.505         4.04         9.64   
16847    16848  506.170  401.173  299.656       628.58       630.88   

       onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  roam_ic_mou_6  \
71941         1.15         41.59         71.13         65.39            0.0   
16847       405.99        365.44        156.06        179.58            0.0   

       roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  \
71941            0.0            0.0            0.0            0.0   
16847            0.0            0.0            0.0            0.0   

       roam_og_mou_8  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
71941            0.0              1.81              5.49              0.00   
16847            0.0              4.13              3.11              1.81   

       loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2f_mou_6  \
71941             33.84             55.96             63.14              0.36   
16847             19.99             25.54             57.33              0.00   

       loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2c_mou_6  loc_og_t2c_mou_7  \
71941               0.0               0.6               0.0               0.0   
16847               0.0               0.0               0.0               0.0   

       loc_og_t2c_mou_8  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  \
71941               0.0         36.03         61.46         63.74   
16847               0.0         24.13         28.66         59.14   

       std_og_t2t_mou_6  std_og_t2t_mou_7  std_og_t2t_mou_8  std_og_t2m_mou_6  \
71941              2.23              4.15              1.15              7.38   
16847            624.44            627.76            404.18            345.44   

       std_og_t2m_mou_7  std_og_t2m_mou_8  std_og_t2f_mou_6  std_og_t2f_mou_7  \
71941             14.99              1.65               0.0               0.0   
16847            130.51            122.24               0.0               0.0   

       std_og_t2f_mou_8  std_og_mou_6  std_og_mou_7  std_og_mou_8  \
71941               0.0          9.61         19.14          2.80   
16847               0.0        969.89        758.28        526.43   

       isd_og_mou_6  isd_og_mou_7  isd_og_mou_8  spl_og_mou_6  spl_og_mou_7  \
71941           0.0          0.16           0.0          0.00          0.00   
16847           0.0          0.00           0.0          4.81          1.68   

       spl_og_mou_8  og_others_6  og_others_7  og_others_8  total_og_mou_6  \
71941           0.0          0.0          0.0          0.0           45.64   
16847           0.0          0.0          0.0          0.0          998.84   

       total_og_mou_7  total_og_mou_8  loc_ic_t2t_mou_6  loc_ic_t2t_mou_7  \
71941           80.78           66.54              0.79              5.21   
16847          788.63          585.58             66.88             40.33   

       loc_ic_t2t_mou_8  loc_ic_t2m_mou_6  loc_ic_t2m_mou_7  loc_ic_t2m_mou_8  \
71941              0.85             42.51             72.01             77.89   
16847              0.11             54.73            170.13            130.66   

       loc_ic_t2f_mou_6  loc_ic_t2f_mou_7  loc_ic_t2f_mou_8  loc_ic_mou_6  \
71941              2.76              5.93             15.89         46.08   
16847              0.00              0.00              0.00        121.61   

       loc_ic_mou_7  loc_ic_mou_8  std_ic_t2t_mou_6  std_ic_t2t_mou_7  \
71941         83.16         94.64              0.00              0.00   
16847        210.46        130.78             24.29             12.39   

       std_ic_t2t_mou_8  std_ic_t2m_mou_6  std_ic_t2m_mou_7  std_ic_t2m_mou_8  \
71941              0.00              0.00              0.00              1.20   
16847             18.28             16.79              2.89             15.28   

       std_ic_t2f_mou_6  std_ic_t2f_mou_7  std_ic_t2f_mou_8  std_ic_mou_6  

# 5 Developing new features

In [43]:
#AON: Age on network 
#Conevrt AON in years
df['aon_yr'] = round((df['aon']/365),1)

In [44]:
df.drop('aon', axis=1, inplace=True)
df.sample(2)

cust_id   arpu_6   arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  \
95781    95782  665.509  601.760  532.533       835.94       748.04   
11744    11745   57.134   70.838   78.482        30.84        15.71   

       onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  roam_ic_mou_6  \
95781       728.21        321.19        277.54        296.64            0.0   
11744        22.13         25.36         49.26         49.43            0.0   

       roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  \
95781            0.0            0.0            0.0            0.0   
11744            0.0            0.0            0.0            0.0   

       roam_og_mou_8  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
95781            0.0            134.98             81.04            108.83   
11744            0.0             30.84             15.71             22.13   

       loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2f_mou_6  \
95781            306.83            238.29            276.43              4.36   
11744             25.36             36.79             49.43              0.00   

       loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2c_mou_6  loc_og_t2c_mou_7  \
95781               3.5             11.54               0.0               0.0   
11744               0.0              0.00               0.0               0.0   

       loc_og_t2c_mou_8  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  \
95781               0.0        446.18        322.84        396.81   
11744               0.0         56.21         52.51         71.56   

       std_og_t2t_mou_6  std_og_t2t_mou_7  std_og_t2t_mou_8  std_og_t2m_mou_6  \
95781            700.96            666.99            619.38              5.98   
11744              0.00              0.00              0.00              0.00   

       std_og_t2m_mou_7  std_og_t2m_mou_8  std_og_t2f_mou_6  std_og_t2f_mou_7  \
95781             35.74              8.66               0.0               0.0   
11744             12.46              0.00               0.0               0.0   

       std_og_t2f_mou_8  std_og_mou_6  std_og_mou_7  std_og_mou_8  \
95781               0.0        706.94        702.74        628.04   
11744               0.0          0.00         12.46          0.00   

       isd_og_mou_6  isd_og_mou_7  isd_og_mou_8  spl_og_mou_6  spl_og_mou_7  \
95781          4.01           0.0           0.0           0.0          0.00   
11744          0.00           0.0           0.0           0.0          4.31   

       spl_og_mou_8  og_others_6  og_others_7  og_others_8  total_og_mou_6  \
95781          0.00          0.0          0.0          0.0         1157.14   
11744          0.26          0.0          0.0          0.0           56.21   

       total_og_mou_7  total_og_mou_8  loc_ic_t2t_mou_6  loc_ic_t2t_mou_7  \
95781         1025.59         1024.86              7.76             33.59   
11744           69.29           71.83              0.76              0.61   

       loc_ic_t2t_mou_8  loc_ic_t2m_mou_6  loc_ic_t2m_mou_7  loc_ic_t2m_mou_8  \
95781             14.71             89.79             73.01            166.41   
11744              9.88             19.63             61.64             22.88   

       loc_ic_t2f_mou_6  loc_ic_t2f_mou_7  loc_ic_t2f_mou_8  loc_ic_mou_6  \
95781              2.41              0.21              0.00         99.98   
11744              0.16              0.00              0.23         20.56   

       loc_ic_mou_7  loc_ic_mou_8  std_ic_t2t_mou_6  std_ic_t2t_mou_7  \
95781        106.83        181.13              5.21             24.08   
11744         62.26         32.99              0.00              0.00   

       std_ic_t2t_mou_8  std_ic_t2m_mou_6  std_ic_t2m_mou_7  std_ic_t2m_mou_8  \
95781             24.56               1.9              5.16              1.69   
11744              0.00               1.8              0.00              0.95   

       std_ic_t2f_mou_6  std_ic_t2f_mou_7  std_ic_t2f_mou_8  std_ic_mou_6  

Let's build bins for the age of network column, which will indicate the number of years a client has been utilizing the T network.

In [45]:
age_range = [ 0,  2,  4,  6,  8, 10, 12]
age_bin = [ 1, 2, 3, 4, 5, 6]
df['age_group'] = pd.cut(df['aon_yr'], age_range, labels=age_bin)
df['age_group'] = df['age_group'].astype(str)
df['age_group'].head()

0    2
1    2
2    2
3    4
4    3
Name: age_group, dtype: object

In [46]:
# let us update our categorical column list
cat_cols.append('age_group')
cat_cols

['night_pck_user_6',
 'night_pck_user_7',
 'night_pck_user_8',
 'fb_user_6',
 'fb_user_7',
 'fb_user_8',
 'age_group']

# 6 Handling Missing values

## 6.1 Checking the null values

In [47]:
null = round(100*(df.isnull().sum()/len(df.index)),2).sort_values(ascending = False)
null = null[null!=0]
null

arpu_2g_6                   74.85
fb_user_6                   74.85
arpu_3g_6                   74.85
date_of_last_rech_data_6    74.85
night_pck_user_6            74.85
total_rech_data_6           74.85
max_rech_data_6             74.85
count_rech_2g_6             74.85
av_rech_amt_data_6          74.85
count_rech_3g_6             74.85
arpu_2g_7                   74.43
count_rech_3g_7             74.43
fb_user_7                   74.43
date_of_last_rech_data_7    74.43
max_rech_data_7             74.43
count_rech_2g_7             74.43
total_rech_data_7           74.43
arpu_3g_7                   74.43
av_rech_amt_data_7          74.43
night_pck_user_7            74.43
count_rech_3g_8             73.66
night_pck_user_8            73.66
arpu_2g_8                   73.66
fb_user_8                   73.66
date_of_last_rech_data_8    73.66
total_rech_data_8           73.66
arpu_3g_8                   73.66
av_rech_amt_data_8          73.66
max_rech_data_8             73.66
count_rech_2g_

`Observation`: After analysing I have got the 74.85% missing values for some features in this dataset.

## 6.2 Imputation of missing values

In [49]:
#Observe missing values in recharge columns 
rech_col = [i for i in df.columns if 'rech' in i]
rech_6_col = [i for i in rech_col if '6' in i]
rech_7_col = [i for i in rech_col if '7' in i]
rech_8_col = [i for i in rech_col if '8' in i]

In [51]:
#Observe missing values in recharge columns in the month of june(6)
rech_6 = pd.DataFrame(df[rech_6_col])

#adding some other columns describing data usage of customer in june(6)
vol_col = df[["vol_2g_mb_6",'vol_3g_mb_6']]

rech_6 = pd.concat([rech_6,vol_col], axis = 1) 
rech_6.sample(2)

total_rech_num_6  total_rech_amt_6  max_rech_amt_6 date_of_last_rech_6  \
46369                 9               110             110          2014-06-27   
78475                 4               100             100          2014-06-27   

      date_of_last_rech_data_6  total_rech_data_6  max_rech_data_6  \
46369               2014-06-24                2.0             98.0   
78475                      NaT                NaN              NaN   

       count_rech_2g_6  count_rech_3g_6  av_rech_amt_data_6  vol_2g_mb_6  \
46369              2.0              0.0               196.0         0.53   
78475              NaN              NaN                 NaN         0.00   

       vol_3g_mb_6  
46369       604.31  
78475         0.00

`Observation`: The missing values for max rech, count rech 2g, and 3g are displayed in the table above whenever the date of the last recharge is absent. As observed in the preceding table, the mobile internet use (2G and 3G data) for these values is zero. As there were no client recharges, I would thus impute these missing values with a value of zero.

In [52]:
#Similarly, observe missing values in recharge columns for the month of july(7)
rech_7_col

['total_rech_num_7',
 'total_rech_amt_7',
 'max_rech_amt_7',
 'date_of_last_rech_7',
 'date_of_last_rech_data_7',
 'total_rech_data_7',
 'max_rech_data_7',
 'count_rech_2g_7',
 'count_rech_3g_7',
 'av_rech_amt_data_7']

In [53]:
rech_7 = pd.DataFrame(df[rech_7_col])

#adding some other columns describing data usage of customer in july
vol_col = df[["vol_2g_mb_7",'vol_3g_mb_7']]

rech_7 = pd.concat([rech_7,vol_col], axis = 1) 
rech_7.sample(2)

total_rech_num_7  total_rech_amt_7  max_rech_amt_7 date_of_last_rech_7  \
68716                12               776             110          2014-07-28   
83408                 3               240             120          2014-07-27   

      date_of_last_rech_data_7  total_rech_data_7  max_rech_data_7  \
68716                      NaT                NaN              NaN   
83408                      NaT                NaN              NaN   

       count_rech_2g_7  count_rech_3g_7  av_rech_amt_data_7  vol_2g_mb_7  \
68716              NaN              NaN                 NaN          0.0   
83408              NaN              NaN                 NaN          0.0   

       vol_3g_mb_7  
68716          0.0  
83408          0.0

`Observation`: When the date of the previous recharge is absent, the associated values for max rech, count rech 2g, and 3g are also missing, as shown in the table above. According to the preceding table, the mobile internet consumption (2G and 3G data) corresponding to these numbers is zero. Therefore, I will replace these missing numbers with zero, as the consumer did not do any recharges. This resembles the trend observed throughout the month of June.

In [54]:
#Similarly,observe missing values in recharge columns in the month of August(8)
rech_8_col

['total_rech_num_8',
 'total_rech_amt_8',
 'max_rech_amt_8',
 'date_of_last_rech_8',
 'date_of_last_rech_data_8',
 'total_rech_data_8',
 'max_rech_data_8',
 'count_rech_2g_8',
 'count_rech_3g_8',
 'av_rech_amt_data_8']

In [55]:
rech_8 = pd.DataFrame(df[rech_8_col])

# adding some other columns describing data usage of customer in August
vol_col = df[["vol_2g_mb_8",'vol_3g_mb_8']]

rech_8 = pd.concat([rech_8,vol_col], axis = 1) 
rech_8.sample(2)

total_rech_num_8  total_rech_amt_8  max_rech_amt_8 date_of_last_rech_8  \
36712                28               777              50          2014-08-31   
27308                16               376              50          2014-08-30   

      date_of_last_rech_data_8  total_rech_data_8  max_rech_data_8  \
36712                      NaT                NaN              NaN   
27308                      NaT                NaN              NaN   

       count_rech_2g_8  count_rech_3g_8  av_rech_amt_data_8  vol_2g_mb_8  \
36712              NaN              NaN                 NaN          0.0   
27308              NaN              NaN                 NaN          0.0   

       vol_3g_mb_8  
36712          0.0  
27308          0.0

`Observation`: Similar to the months of June and July, the values for max rech, count rech 2g, and 3g are missing if the date of the last recharge is absent. According to the preceding table, the mobile internet consumption (2G and 3G data) corresponding to these numbers is zero. Therefore, let's impute these missing numbers with zero, given that there were no client recharges. This is identical to the trend observed throughout the months of June and July.

## 6.3 Filling the missing value by zero: Numerical features

In [57]:
# Let's use zero for the missing recharge column data.

impute_0 = [ 'date_of_last_rech_data_6','max_rech_data_6','count_rech_2g_6','count_rech_3g_6',
           'date_of_last_rech_data_7','max_rech_data_7','count_rech_2g_7','count_rech_3g_7',
           'date_of_last_rech_data_8','max_rech_data_8','count_rech_2g_8','count_rech_3g_8']

df[impute_0] = df[impute_0].apply(lambda x: x.fillna(0))

In [58]:
# Now Checking Null values
null = round(100*(df.isnull().sum()/len(df.index)),2).sort_values(ascending = False)
null = null[null!=0]
null

fb_user_6              74.85
arpu_2g_6              74.85
total_rech_data_6      74.85
av_rech_amt_data_6     74.85
arpu_3g_6              74.85
night_pck_user_6       74.85
fb_user_7              74.43
total_rech_data_7      74.43
arpu_2g_7              74.43
night_pck_user_7       74.43
arpu_3g_7              74.43
av_rech_amt_data_7     74.43
arpu_3g_8              73.66
arpu_2g_8              73.66
night_pck_user_8       73.66
fb_user_8              73.66
total_rech_data_8      73.66
av_rech_amt_data_8     73.66
std_og_t2m_mou_8        5.38
std_og_mou_8            5.38
std_og_t2f_mou_8        5.38
std_ic_t2m_mou_8        5.38
std_ic_mou_8            5.38
std_ic_t2f_mou_8        5.38
loc_ic_t2f_mou_8        5.38
std_ic_t2t_mou_8        5.38
isd_og_mou_8            5.38
std_og_t2t_mou_8        5.38
loc_ic_mou_8            5.38
spl_og_mou_8            5.38
og_others_8             5.38
loc_ic_t2m_mou_8        5.38
spl_ic_mou_8            5.38
loc_og_t2c_mou_8        5.38
loc_og_t2m_mou

## 6.4 Imputing Categorical columns

In [59]:
#Impute the missing values in categorical columns by (-1). 
df[cat_cols] = df[cat_cols].apply(lambda x: x.fillna(-1)) 
df[cat_cols] = df[cat_cols].astype('str')
df[cat_cols].sample(2)

night_pck_user_6 night_pck_user_7 night_pck_user_8 fb_user_6 fb_user_7  \
76580             -1.0             -1.0              0.0      -1.0      -1.0   
55282             -1.0             -1.0             -1.0      -1.0      -1.0   

      fb_user_8 age_group  
76580       0.0         1  
55282      -1.0         5

In [60]:
# Now Checking Null values
null = round(100*(high_val.isnull().sum()/len(high_val.index)),2).sort_values(ascending = False)
null = null[null!=0]
null

arpu_2g_6                   64.05
date_of_last_rech_data_6    64.05
count_rech_3g_6             64.05
count_rech_2g_6             64.05
max_rech_data_6             64.05
night_pck_user_6            64.05
total_rech_data_6           64.05
arpu_3g_6                   64.05
fb_user_6                   64.05
av_rech_amt_data_6          64.05
date_of_last_rech_data_7    62.29
av_rech_amt_data_7          62.29
max_rech_data_7             62.29
fb_user_7                   62.29
count_rech_2g_7             62.29
arpu_2g_7                   62.29
count_rech_3g_7             62.29
total_rech_data_7           62.29
night_pck_user_7            62.29
arpu_3g_7                   62.29
arpu_3g_8                   59.43
fb_user_8                   59.43
count_rech_3g_8             59.43
av_rech_amt_data_8          59.43
count_rech_2g_8             59.43
arpu_2g_8                   59.43
date_of_last_rech_data_8    59.43
max_rech_data_8             59.43
total_rech_data_8           59.43
night_pck_user

## 6.5 Dropping columns with more than 40% missing values

More than forty percent of the average revenue per user during both the good and action phases are missing. Since the focus of this research is the identification of consumers who are likely to churn based only on usage-based churn, I may safely exclude these columns without compromising the research.

In [62]:
miss = round(100*(df.isnull().sum()/len(df.index)),2).sort_values(ascending = False) 

miss = pd.DataFrame(miss[miss >= 40])
threshold_col = miss.index

df = df.drop(threshold_col,1)

In [63]:
df.shape

(99999, 152)

## 6.6 Dropping missing values for out going calls for month of June, July, and August

#### For Month of June

In [66]:
#According to my business domain knowledge, the total_og column represents the sum of local, standard, special, and other outgoing calls.
og_call_6 = ['loc_og_mou_6','std_og_mou_6','isd_og_mou_6','spl_og_mou_6','og_others_6','total_og_mou_6']
total_og_6 = df[og_call_6]

#limiting the filtering to clients who have made no outbound calls.
total_og_6.loc[total_og_6['total_og_mou_6']==0].sample(2)

loc_og_mou_6  std_og_mou_6  isd_og_mou_6  spl_og_mou_6  og_others_6  \
26815           0.0           0.0           0.0           0.0          0.0   
37108           NaN           NaN           NaN           NaN          NaN   

       total_og_mou_6  
26815             0.0  
37108             0.0

`Observation`: Based on my understanding of the business area and the preceding table, I can deduce that when the total number of outgoing calls for a given month is 0, no outgoing calls have been made. I am going to change the missing values for local outgoing calls, normal outgoing calls, special outgoing calls, and other outgoing calls with 0.

In [67]:
#Let's see whether our assumption is correct or not: the sum of all outgoing calls equals the 'total_og_mou_6'.

df['outgoing_total_6'] = df['loc_og_mou_6']+ df['std_og_mou_6']+df['isd_og_mou_6']+df['spl_og_mou_6']+df['og_others_6']
df[['outgoing_total_6','total_og_mou_6']].dropna().corr()

outgoing_total_6  total_og_mou_6
outgoing_total_6               1.0             1.0
total_og_mou_6                 1.0             1.0

`Observation`: From the above table, it is evident that these variables are associated; thus, when the total number of outgoing calls for a given month is 0,that means no outbound calls have been made. as a result, I am going to replace the missing values for 'local outgoing calls', 'normal outgoing calls', 'special outgoing calls', and 'other outgoing calls' with 0.

In [69]:
#Imputing by '0'

df[og_call_6] = df[og_call_6].fillna(0)

### local outgoing calls(T2T,T2M,T2F & T2C)

In [70]:
og_loc_6 = ['loc_og_t2t_mou_6','loc_og_t2m_mou_6','loc_og_t2f_mou_6','loc_og_t2c_mou_6','loc_og_mou_6']
loc_og_6 = df[og_loc_6]

# filtering only those clients who didn't make any 'local outgoing calls'
loc_og_6.loc[loc_og_6 ['loc_og_mou_6']==0].sample(2)

loc_og_t2t_mou_6  loc_og_t2m_mou_6  loc_og_t2f_mou_6  loc_og_t2c_mou_6  \
78032               0.0               0.0               0.0               0.0   
9041                0.0               0.0               0.0               0.0   

       loc_og_mou_6  
78032           0.0  
9041            0.0

Local outbound calls in any given month are equal to the total of all local outgoing call kinds, i.e. (T2T,T2M,T2F & T2C)

The monthly total number of local outbound calls (T2T,T2M,T2F, and T2C) is zero. Now, I will replace the missing value with zero.

In [71]:
#Lets verify is the assumption right or not, sum of 'all local outgoing calls' is 
#equal to the the 'total local outgoing call'
#"loc_og_mou_6"

df['loc_outgoing_total_6'] = df['loc_og_t2t_mou_6']+ df['loc_og_t2m_mou_6']+df['loc_og_t2f_mou_6']+df['loc_og_t2c_mou_6']
df[['loc_outgoing_total_6','loc_og_mou_6']].dropna().corr().round()

loc_outgoing_total_6  loc_og_mou_6
loc_outgoing_total_6                   1.0           1.0
loc_og_mou_6                           1.0           1.0

`Observation:` From the above table, it is evident that these variables are associated; thus, when the total number of local outgoing calls for a given month is 0,that means no local outbound calls have been made. as a result, I am going to replace the missing values for T2T,T2M,T2F & T2C by zero

In [72]:
# Imputing by '0'
df[og_loc_6] = df[og_loc_6].fillna(0)

#### total standard outgoing calls 

In [74]:
og_std_6 = ['std_og_t2t_mou_6','std_og_t2m_mou_6','std_og_t2f_mou_6','std_og_mou_6']
std_og_6 = df[og_std_6]
std_og_6.sample(5)

std_og_t2t_mou_6  std_og_t2m_mou_6  std_og_t2f_mou_6  std_og_mou_6
39992             14.93              0.46              0.00         15.39
49298            177.36             99.61              0.00        276.98
72895              0.00              0.00              0.66          0.66
68615             19.44             15.73             14.49         49.68
16297            217.08              5.49              0.00        222.58

In [75]:
# Imputing by '0'
df[og_std_6] = df[og_std_6].fillna(0)

#### For Month of July

Similar process is applied like June

#### Total Outgoing Call

In [77]:
# From the business domain knowledge, we know that total_og column is the addition of local , standard, special and other outgoing calls
og_call_7 = ['loc_og_mou_7','std_og_mou_7','isd_og_mou_7','spl_og_mou_7','og_others_7','total_og_mou_7']
total_og_7 = high_val[og_call_7]
# filtering only those clients who have made no outgoing calls
total_og_7.loc[total_og_7['total_og_mou_7']==0].sample(2)

loc_og_mou_7  std_og_mou_7  isd_og_mou_7  spl_og_mou_7  og_others_7  \
91223           NaN           NaN           NaN           NaN          NaN   
58319           0.0           0.0           0.0           0.0          0.0   

       total_og_mou_7  
91223             0.0  
58319             0.0

In [78]:
#Lets verify is our assumption right or not, sum of all outgoing calls is equal to the the tota_og_mou_7
df['outgoing_total_7'] = df['loc_og_mou_7']+ df['std_og_mou_7']+df['isd_og_mou_7']+df['spl_og_mou_7']+df['og_others_7']
df[['outgoing_total_7','total_og_mou_7']].dropna().corr()

outgoing_total_7  total_og_mou_7
outgoing_total_7               1.0             1.0
total_og_mou_7                 1.0             1.0

In [79]:
# Imputing by '0'
df[og_call_7] = df[og_call_7].fillna(0)

In [81]:
og_loc_7 = ['loc_og_t2t_mou_7','loc_og_t2m_mou_7','loc_og_t2f_mou_7','loc_og_t2c_mou_7','loc_og_mou_7']
loc_og_7 = high_val[og_loc_7]

# filtering only those clients who didn't made any local outgoing calls
loc_og_7.loc[loc_og_7 ['loc_og_mou_7']==0].sample(3)

loc_og_t2t_mou_7  loc_og_t2m_mou_7  loc_og_t2f_mou_7  loc_og_t2c_mou_7  \
77218               0.0               0.0               0.0              0.00   
93697               0.0               0.0               0.0              0.01   
35221               0.0               0.0               0.0              0.00   

       loc_og_mou_7  
77218           0.0  
93697           0.0  
35221           0.0

#### T2T,T2M,T2F & T2C

In [82]:
#Lets verify is our assumption right or not, sum of all local outgoing calls is equal to 
#the total local outgoing call
#"loc_og_mou_7"

df['loc_outgoing_total_7'] = df['loc_og_t2t_mou_7']+ df['loc_og_t2m_mou_7']+df['loc_og_t2f_mou_7']+df['loc_og_t2c_mou_7']
df[['loc_outgoing_total_7','loc_og_mou_7']].dropna().corr().round()

loc_outgoing_total_7  loc_og_mou_7
loc_outgoing_total_7                   1.0           1.0
loc_og_mou_7                           1.0           1.0

In [83]:
# Imputing by '0'
df[og_loc_7] = df[og_loc_7].fillna(0)

In [84]:
og_std_7 = ['std_og_t2t_mou_7','std_og_t2m_mou_7','std_og_t2f_mou_7','std_og_mou_7']
std_og_7 = high_val[og_std_7]
std_og_7.sample(3)

std_og_t2t_mou_7  std_og_t2m_mou_7  std_og_t2f_mou_7  std_og_mou_7
83968           1428.46             71.19              0.00       1499.66
89162             15.38             33.51             12.51         61.41
56651            541.61            229.96              0.00        771.58

#### Total standard outgoing calls 

In [85]:
#Lets verify is our assumption right or not, sum of all std outgoing calls is equal to 
#the total std outgoing call
#"std_og_mou_7"

df['std_outgoing_total_7'] = df['std_og_t2t_mou_7']+ df['std_og_t2m_mou_7']+df['std_og_t2f_mou_7']
df[['std_outgoing_total_7','std_og_mou_7']].dropna().corr()

std_outgoing_total_7  std_og_mou_7
std_outgoing_total_7                   1.0           1.0
std_og_mou_7                           1.0           1.0

In [86]:
# Imputing by '0'
df[og_std_7] = df[og_std_7].fillna(0)

#### For Month of August

Similar process is applied like June

#### Total Outgoing Call

In [88]:
# From the business domain knowledge, we know that total_og column is the addition of local , standard, special and other outgoing calls
og_call_8 = ['loc_og_mou_8','std_og_mou_8','isd_og_mou_8','spl_og_mou_8','og_others_8','total_og_mou_8']
total_og_8 = high_val[og_call_8]
# filtering only those clients who have made no outgoing calls
total_og_8.loc[total_og_8['total_og_mou_8']==0].sample(3)

loc_og_mou_8  std_og_mou_8  isd_og_mou_8  spl_og_mou_8  og_others_8  \
24362           0.0           0.0           0.0           0.0          0.0   
69272           0.0           0.0           0.0           0.0          0.0   
80071           0.0           0.0           0.0           0.0          0.0   

       total_og_mou_8  
24362             0.0  
69272             0.0  
80071             0.0

In [89]:
#Lets verify is our assumption right or not, sum of 
#all outgoing calls is equal to the tota_og_mou_8

df['outgoing_total_8'] = df['loc_og_mou_8']+ df['std_og_mou_8']+df['isd_og_mou_8']+df['spl_og_mou_8']+df['og_others_8']
df[['outgoing_total_8','total_og_mou_8']].dropna().corr()

outgoing_total_8  total_og_mou_8
outgoing_total_8               1.0             1.0
total_og_mou_8                 1.0             1.0

In [90]:
# Imputing by '0'
df[og_call_8] = df[og_call_8].fillna(0)

#### T2T,T2M,T2F & T2C

In [92]:
og_loc_8 = ['loc_og_t2t_mou_8','loc_og_t2m_mou_8','loc_og_t2f_mou_8','loc_og_t2c_mou_8','loc_og_mou_8']
loc_og_8 = high_val[og_loc_8]

# filtering only those clients who have made no local outgoing calls
loc_og_8.loc[loc_og_8 ['loc_og_mou_8']==0].sample(2)

loc_og_t2t_mou_8  loc_og_t2m_mou_8  loc_og_t2f_mou_8  loc_og_t2c_mou_8  \
58810               0.0               0.0               0.0              4.01   
85655               0.0               0.0               0.0              0.00   

       loc_og_mou_8  
58810           0.0  
85655           0.0

In [93]:
#Lets verify is our assumption right or not, sum of all local outgoing calls is equal to the the total local outgoing call
#"loc_og_mou_8"

df['loc_outgoing_total_8'] = df['loc_og_t2t_mou_8']+ df['loc_og_t2m_mou_8']+df['loc_og_t2f_mou_8']+df['loc_og_t2c_mou_8']
df[['loc_outgoing_total_8','loc_og_mou_8']].dropna().corr().round()

loc_outgoing_total_8  loc_og_mou_8
loc_outgoing_total_8                   1.0           1.0
loc_og_mou_8                           1.0           1.0

In [94]:
# Imputing by '0'
df[og_loc_8] = df[og_loc_8].fillna(0)

#### std outgoing calls

In [96]:
og_std_8 = ['std_og_t2t_mou_8','std_og_t2m_mou_8','std_og_t2f_mou_8','std_og_mou_8']
std_og_8 = df[og_std_8]
std_og_8.sample(3)

std_og_t2t_mou_8  std_og_t2m_mou_8  std_og_t2f_mou_8  std_og_mou_8
68119              0.00              0.00               0.0          0.00
55018              1.08             36.39               0.0         37.48
79577              0.00              0.00               0.0          0.00

In [97]:
#Lets verify is our assumption right or not, sum of all std outgoing 
#calls is equal to the total std outgoing call
#"std_og_mou_8"

df['std_outgoing_total_8'] = df['std_og_t2t_mou_8']+ df['std_og_t2m_mou_8']+df['std_og_t2f_mou_8']
df[['std_outgoing_total_8','std_og_mou_8']].dropna().corr()

std_outgoing_total_8  std_og_mou_8
std_outgoing_total_8                   1.0           1.0
std_og_mou_8                           1.0           1.0

In [98]:
# Imputing by '0'
high_val[og_std_8] = high_val[og_std_8].fillna(0)

## 6.7 Dropping missing values for incoming calls for month of June, July, and August

usage for local, standard, special and others¶

#### For month of June

In [99]:
# From the business domain knowledge, we know that total_ic column is the addition of local , standard, special andd other incoming calls
ic_call_6 = ['loc_ic_mou_6','std_ic_mou_6','isd_ic_mou_6','spl_ic_mou_6','ic_others_6','total_ic_mou_6']
total_ic_6 = df[ic_call_6]
# filtering only those clients who have made no outgoing calls
total_ic_6 .loc[total_ic_6 ['total_ic_mou_6']==0].sample(2)

loc_ic_mou_6  std_ic_mou_6  isd_ic_mou_6  spl_ic_mou_6  ic_others_6  \
14986           NaN           NaN           NaN           NaN          NaN   
92807           0.0           0.0           0.0           0.0          0.0   

       total_ic_mou_6  
14986             0.0  
92807             0.0

`Observation`: Based on my understanding of the business domain and the aforementioned statistics, I can deduce that when the total number of incoming calls is zero, no incoming calls are received during that month. The missing values can be replaced with 0 in the columns for local incoming calls, regular incoming calls, special incoming calls, and other incoming calls.

In [100]:
#Lets verify is our assumption right or not, sum of all incoming calls is equal to the tota_ic_mou_6

df['incoming_total_6'] = df['loc_ic_mou_6']+ df['std_ic_mou_6']+df['isd_ic_mou_6']+df['spl_ic_mou_6']+df['ic_others_6']
df[['incoming_total_6','total_ic_mou_6']].dropna().corr()

incoming_total_6  total_ic_mou_6
incoming_total_6               1.0             1.0
total_ic_mou_6                 1.0             1.0

`Observation`: As can be seen in the table above, when the total number of incoming calls is 0, no outgoing calls have been made during that month. 0 can be substituted for missing values in the columns for local outgoing calls, normal outgoing calls, special incoming calls, and other outgoing calls.

In [101]:
# Imputing by '0'
df[ic_call_6] = df[ic_call_6].fillna(0)

#### T2T,T2M,T2F & T2C

In [103]:
ic_loc_6 = ['loc_ic_t2t_mou_6','loc_ic_t2m_mou_6','loc_ic_t2f_mou_6','loc_ic_mou_6']
loc_ic_6= high_val[ic_loc_6]

# filtering only those clients who have made no local outgoing calls
loc_ic_6.loc[loc_ic_6['loc_ic_mou_6']==0].sample(3)

loc_ic_t2t_mou_6  loc_ic_t2m_mou_6  loc_ic_t2f_mou_6  loc_ic_mou_6
90772               0.0               0.0               0.0           0.0
91056               0.0               0.0               0.0           0.0
99875               0.0               0.0               0.0           0.0

`Observation`:If the total number of local calls received in a particular month is zero, no local calls were received, i.e. T2T, T2M, T2F, and T2C. I am able to replace '0' for the missing equivalent numbers.

In [104]:
#Lets verify is our assumption right or not, sum of all local outgoing calls is equal to the 
#total local outgoing call
#"loc_ic_mou_6"

df['loc_incoming_total_6'] = df['loc_ic_t2t_mou_6']+ df['loc_ic_t2m_mou_6']+df['loc_ic_t2f_mou_6']
df[['loc_incoming_total_6','loc_ic_mou_6']].dropna().corr()

loc_incoming_total_6  loc_ic_mou_6
loc_incoming_total_6                   1.0           1.0
loc_ic_mou_6                           1.0           1.0

In [105]:
# Imputing by '0'
df[ic_loc_6] = df[ic_loc_6].fillna(0)

#### std incoming calls

It is obvious from the statistics that standard incoming calls in any given month equal the total of all standard incoming call categories, i.e. (T2T,T2M,T2F & T2C)

In [107]:
ic_std_6 = ['std_ic_t2t_mou_6','std_ic_t2m_mou_6','std_ic_t2f_mou_6','std_ic_mou_6']
std_ic_6 = df[ic_std_6]
std_ic_6.sample(3)

std_ic_t2t_mou_6  std_ic_t2m_mou_6  std_ic_t2f_mou_6  std_ic_mou_6
32871               0.0              0.00              0.58          0.58
82174               0.0              8.11              0.00          8.11
67676               0.0              0.00              1.85          1.85

`Observation`:If the total number of regular outbound calls in a particular month is zero, no T2T, T2M, T2F, or T2C calls were made. I am able to substitute '0' for the missing values that correspond to them.

In [108]:
#Lets verify is our assumption right or not, sum of all std outgoing calls is equal to the total std outgoing call
#"std_ic_mou_6"

df['std_incoming_total_6'] = df['std_ic_t2t_mou_6']+ df['std_ic_t2m_mou_6']+df['std_ic_t2f_mou_6']
df[['std_incoming_total_6','std_ic_mou_6']].dropna().corr()

std_incoming_total_6  std_ic_mou_6
std_incoming_total_6                   1.0           1.0
std_ic_mou_6                           1.0           1.0

`Observation`:If the total number of standard outbound calls in a particular month is 0, then no standard calls, i.e. T2T, T2M, T2F, and T2C, were made.

In [109]:
# Imputing by '0'
df[ic_std_6] = df[ic_std_6].fillna(0)

#### For month of July

# Total Calls

In [111]:
# From the business domain knowledge, we know that total_ic column is the addition of local , standard, special andd other incoming calls
ic_call_7 = ['loc_ic_mou_7','std_ic_mou_7','isd_ic_mou_7','spl_ic_mou_7','ic_others_7','total_ic_mou_7']
total_ic_7 = high_val[ic_call_7]
# filtering only those clients who have made no outgoing calls
total_ic_7 .loc[total_ic_7 ['total_ic_mou_7']==0].sample(2)

loc_ic_mou_7  std_ic_mou_7  isd_ic_mou_7  spl_ic_mou_7  ic_others_7  \
4916            0.0           0.0           0.0           0.0          0.0   
54195           0.0           0.0           0.0           0.0          0.0   

       total_ic_mou_7  
4916              0.0  
54195             0.0

In [112]:
#Lets verify is our assumption right or not, sum of all local incoming calls is equal to the the total local incoming call
#"loc_ic_mou_7"

df['loc_incoming_total_7'] = df['loc_ic_mou_7']+ df['std_ic_mou_7']+df['isd_ic_mou_7']+df['spl_ic_mou_7']+df['ic_others_7']
df[['loc_incoming_total_7','total_ic_mou_7']].dropna().corr()

loc_incoming_total_7  total_ic_mou_7
loc_incoming_total_7                   1.0             1.0
total_ic_mou_7                         1.0             1.0

In [113]:
# Imputing by '0'
high_val[ic_call_7] = high_val[ic_call_7].fillna(0)

#### T2T,T2M,T2F & T2C

In [115]:
ic_loc_7 = ['loc_ic_t2t_mou_7','loc_ic_t2m_mou_7','loc_ic_t2f_mou_7','loc_ic_mou_7']
loc_ic_7= df[ic_loc_7]

# filtering only those clients who have made no local outgoing calls
loc_ic_7.loc[loc_ic_7['loc_ic_mou_7']==0].sample(3)

loc_ic_t2t_mou_7  loc_ic_t2m_mou_7  loc_ic_t2f_mou_7  loc_ic_mou_7
86206               0.0               0.0               0.0           0.0
49820               0.0               0.0               0.0           0.0
15333               0.0               0.0               0.0           0.0

In [116]:
#Lets verify is our assumption right or not, sum of all local incoming calls is equal to the 
#total local incoming call
#"loc_ic_mou_7"

df['loc_incoming_total_7'] = df['loc_ic_t2t_mou_7']+ df['loc_ic_t2m_mou_7']+df['loc_ic_t2f_mou_7']
df[['loc_incoming_total_7','loc_ic_mou_7']].dropna().corr()

loc_incoming_total_7  loc_ic_mou_7
loc_incoming_total_7                   1.0           1.0
loc_ic_mou_7                           1.0           1.0

In [117]:
# Imputing by '0'
high_val[ic_loc_7] = high_val[ic_loc_7].fillna(0)

#### Standard Outgoing calls

In [119]:
ic_std_7 = ['std_ic_t2t_mou_7','std_ic_t2m_mou_7','std_ic_t2f_mou_7','std_ic_mou_7']
std_ic_7 = df[ic_std_7]
std_ic_7.sample(2)

std_ic_t2t_mou_7  std_ic_t2m_mou_7  std_ic_t2f_mou_7  std_ic_mou_7
95519               0.0               0.0               0.0           0.0
27967               0.0               0.6               0.0           0.6

In [120]:
#Lets verify is our assumption right or not, sum of all local incoming calls is equal to the the total local incoming call
#"std_ic_mou_7"

df['std_incoming_total_7'] = df['std_ic_t2t_mou_7']+ df['std_ic_t2m_mou_7']+df['std_ic_t2f_mou_7']
df[['std_incoming_total_7','std_ic_mou_7']].dropna().corr()

std_incoming_total_7  std_ic_mou_7
std_incoming_total_7                   1.0           1.0
std_ic_mou_7                           1.0           1.0

In [121]:
# Imputing by '0'
high_val[ic_std_7] = high_val[ic_std_7].fillna(0)

#### For August Month

In [123]:
# From the business domain knowledge, we know that total_ic column is the addition of local , standard, special andd other incoming calls
ic_call_8 = ['loc_ic_mou_8','std_ic_mou_8','isd_ic_mou_8','spl_ic_mou_8','ic_others_8','total_ic_mou_8']
total_ic_8 = df[ic_call_8]
# filtering only those clients who have made no outgoing calls
total_ic_8 .loc[total_ic_8['total_ic_mou_8']==0].sample(3)

loc_ic_mou_8  std_ic_mou_8  isd_ic_mou_8  spl_ic_mou_8  ic_others_8  \
4814            0.0           0.0           0.0           0.0          0.0   
51257           NaN           NaN           NaN           NaN          NaN   
81760           0.0           0.0           0.0           0.0          0.0   

       total_ic_mou_8  
4814              0.0  
51257             0.0  
81760             0.0

In [124]:
#Lets verify is our assumption right or not, sum of all incoming calls is equal to the the tota_ic_mou_8

df['incoming_total_8'] = df['loc_ic_mou_8']+ df['std_ic_mou_8']+df['isd_ic_mou_8']+df['spl_ic_mou_8']+df['ic_others_8']
df[['incoming_total_8','total_ic_mou_8']].dropna().corr()

incoming_total_8  total_ic_mou_8
incoming_total_8               1.0             1.0
total_ic_mou_8                 1.0             1.0

In [125]:
# Imputing by '0'
high_val[ic_call_8] = high_val[ic_call_8].fillna(0)

In [127]:
ic_loc_8 = ['loc_ic_t2t_mou_8','loc_ic_t2m_mou_8','loc_ic_t2f_mou_8','loc_ic_mou_8']
loc_ic_8= df[ic_loc_8]

# filtering only those clients who have made no local outgoing calls
loc_ic_8.loc[loc_ic_8['loc_ic_mou_8']==0].sample(3)

loc_ic_t2t_mou_8  loc_ic_t2m_mou_8  loc_ic_t2f_mou_8  loc_ic_mou_8
55959               0.0               0.0               0.0           0.0
89241               0.0               0.0               0.0           0.0
46235               0.0               0.0               0.0           0.0

In [128]:
#Lets verify is our assumption right or not, sum of all local incoming calls is equal to the the total local incoming call
#"loc_ic_mou_8"

df['loc_incoming_total_8'] = df['loc_ic_t2t_mou_8']+ df['loc_ic_t2m_mou_8']+df['loc_ic_t2f_mou_8']
df[['loc_incoming_total_8','loc_ic_mou_8']].dropna().corr()

loc_incoming_total_8  loc_ic_mou_8
loc_incoming_total_8                   1.0           1.0
loc_ic_mou_8                           1.0           1.0

In [129]:
# Imputing by '0'
high_val[ic_loc_8] = high_val[ic_loc_8].fillna(0)

In [131]:
ic_std_8 = ['std_ic_t2t_mou_8','std_ic_t2m_mou_8','std_ic_t2f_mou_8','std_ic_mou_8']
std_ic_8 = df[ic_std_8]
std_ic_8.sample(3)

std_ic_t2t_mou_8  std_ic_t2m_mou_8  std_ic_t2f_mou_8  std_ic_mou_8
84878              1.00              0.00              0.00          1.00
59047              0.00              9.23              0.00          9.23
16165              0.71             88.06              7.01         95.79

In [132]:
#Lets verify is our assumption right or not, sum of all local incoming calls is equal to the the total local incoming call
#"std_ic_mou_8"

df['std_incoming_total_8'] = df['std_ic_t2t_mou_8']+ df['std_ic_t2m_mou_8']+df['std_ic_t2f_mou_8']
df[['std_incoming_total_8','std_ic_mou_8']].dropna().corr()

std_incoming_total_8  std_ic_mou_8
std_incoming_total_8                   1.0           1.0
std_ic_mou_8                           1.0           1.0

In [133]:
# Imputing by '0'
df[ic_std_8] = df[ic_std_8].fillna(0)

In [ ]:
## Let us now check the % of null values
null = round(100*(high_val.isnull().sum()/len(high_val.index)),2).sort_values(ascending = False)
null = null[null!=0]
null